### PR TITLE
Felipemarkson/issue221

### LINKED ISSUE #221: `Alpine` have an issue with `@NLconstraint` created from a quadratic `@expression`
_NOTE: This issue was created after the changes indicated on #220._

When you try to create a`@NLconstraint` quadratic constraint from a `@expression`, `Alpine` throws an error.

The issue is not found if using only `Pavito` to solve the problem, so I think the problem is not the MIP solver.

The issue is not found if the `@expression` is linear.

```julia
using JuMP
using Ipopt, Alpine, HiGHS, Pavito

# MILP optimizer
highs = optimizer_with_attributes(HiGHS.Optimizer,
    MOI.Silent() => true
)

# NLP optimizer
ipopt = optimizer_with_attributes(Ipopt.Optimizer,
    MOI.Silent() => true
)

# Convex MIP optimizer
pavito = optimizer_with_attributes(Pavito.Optimizer,
    "cont_solver" => ipopt,
    "mip_solver" => highs
)

# Global optimizer
alpine = optimizer_with_attributes(Alpine.Optimizer,
    "nlp_solver" => ipopt,
    "mip_solver" => pavito,
    "presolve_bt" => true,
    "apply_partitioning" => true,
)

model = Model(alpine)
@variable(model, x ≥ 0)
@expression(model, expr, x^2)
@NLconstraint(model, expr ≥ 3)
@NLobjective(model, Min, x^3)

optimize!(model)
```

Here you can see the Error message:
image

### PR DESCRIPTION
**Pull request by felipemarkson**
This PR fixes #221.

The problem in #221 was an evaluation in `expr_resolve_const` of an empty sum expression `+()` that JuMP adds by default.

A new function called `expr_is_emptysum` is added to the `Alpine` to check if the expression is an invalid empty sum `+()` that couldn't be evaluated.

If the expression is an empty sum, the `expr_resolve_const` substitute this with a zero-equivalent expression `:(0)`.

This PR also adds unit tests for the new `expr_is_emptysum` and a test for the example in #221.

_NOTE: There could be a better strategy to identify an empty sum expression. But this PR solves the problem for now._

cc @Ibrahim-a-Ibrahim

### CODE REVIEW

File: src/nlexpr.jl [hunk 1]
```diff
@@ -572,6 +572,22 @@ function expr_arrangeargs(args::Array; kwargs...)
     return
 end
 
+"""
+Check if it is an sample of `+()` (https://github.com/lanl-ansi/Alpine.jl/issues/221)
+"""
+function expr_is_emptysum(expr)
+    try
+        _ = eval(expr)
```
Comment (odow) [inline src/nlexpr.jl:580]: Using `eval` is a strong signal that you're doing something wrong 😆 

If I understand correctly, all you need is:
```julia
expr_is_empty_sum(expr) = expr == :(+())
```

Comment (felipemarkson) [reply src/nlexpr.jl:580]: Thanks, @odow!

I'm not very familiar with Julia's expressions 😄

Comment (felipemarkson) [reply src/nlexpr.jl:580]: @odow BTW, why JuMP is adding this empty expression?

Comment (odow) [reply src/nlexpr.jl]: I'm not sure. I'll take a look:
```julia
julia> model = Model();

julia> @variable(model, x);

julia> @expression(model, expr, x^2);

julia> @NLconstraint(model, expr >= 3)
(+() + x * x) - 3.0 ≥ 0
```

**Comment by odow** [comment]
I have a fix for JuMP: https://github.com/jump-dev/JuMP.jl/pull/3091, but we should still implement this work-around in Alpine for now.

**Comment by felipemarkson** [comment]
This should take this amount of time to finished the CI?

It is taking 5 hours...

**Comment by harshangrjn** [comment]
Merged into #224

**Pull request closed without merging**